File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/ReadonlyTransaction.scala
package dev.typr.foundationssc.docs.core

import dev.typr.foundationssc.*

@SuppressWarnings(Array("unused"))
object ReadonlyTransaction:
  case class User(name: String, email: String)

  val userCodec: RowCodecNamed[User] = RowCodec
    .namedBuilder[User]()
    .field("name", PgTypes.text)(_.name)
    .field("email", PgTypes.text)(_.email)
    .build(User.apply)

  private val tx: Transactor = null // placeholder

  val findAll: OperationRead[List[User]] =
    Fragment.of("SELECT name, email FROM users").query(userCodec.all())

  // start
  // Single read operation — no transaction overhead
  def allUsers(): List[User] =
    findAll.transactRead(tx)
  // stop

File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/ReadonlyTransaction.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.*;
import java.util.List;

@SuppressWarnings("unused")
public class ReadonlyTransaction {
  record User(String name, String email) {}

  static RowCodec<User> userCodec =
      RowCodec.<User>builder()
          .field(PgTypes.text, User::name)
          .field(PgTypes.text, User::email)
          .build(User::new);

  Transactor tx = null; // placeholder

  OperationRead<List<User>> findAll =
      Fragment.of("SELECT name, email FROM users").query(userCodec.all());

  //start
  // Single read operation — no transaction overhead
  List<User> allUsers() {
    return findAll.transactRead(tx);
  }
  //stop
}
